drivers/timer/cortex_m_systick.c

```diff
@@ -284,6 +351,47 @@ uint64_t sys_clock_cycle_get_64(void)
 
 void sys_clock_idle_exit(void)
 {
+#ifdef CONFIG_CORTEX_M_SYSTICK_IDLE_TIMER
+	if (timeout_idle) {
+		k_spinlock_key_t key = k_spin_lock(&idle_lock);
+		uint32_t idle_timer_diff;
+		uint32_t idle_timer_post;
+		cycle_t systick_diff;
+		cycle_t missed_cycles;
+		uint64_t systick_us;
+		uint64_t idle_timer_us;
+		uint64_t measurement_diff_us;
+
+		/* Get current values for both timers */
+		counter_get_value(idle_timer, &idle_timer_post);
+		systick_diff = cycle_count + elapsed() - cycle_pre_idle;
+
+		/* Calculate has much time has pasted since last measurement for both timers */
+		idle_timer_diff = idle_timer_post - idle_timer_pre_idle;
+		idle_timer_us = counter_ticks_to_us(idle_timer, idle_timer_diff);
+		systick_us =
+			((uint64_t)systick_diff * USEC_PER_SEC) / sys_clock_hw_cycles_per_sec();
+
+		/* Calculate difference in measurements to get how much time
+		 * the SysTick missed in idle state.
+		 */
+		measurement_diff_us = idle_timer_us - systick_us;
+		missed_cycles =
+			(sys_clock_hw_cycles_per_sec() * measurement_diff_us) / USEC_PER_SEC;
+
+		/* Update the cycle counter to include the cycles missed in idle */
+		cycle_count += missed_cycles;
+
+		/* We've alredy performed all needed operations */
+		timeout_idle = false;
+
+		/* Request sys_clock_announce ASAP */
+		sys_clock_set_timeout(0, false);
```

Comment:
Can't you call `sys_clock_announce()` at that point ?

Reply:
It’s a very good point. I’ve been considering two solutions. I feel like the current solution is safer, the announcement is in one place, code is not repeated, cycles to announce are aligned to tick boundary. However it introduces an additional delay. According to my measurement, delay between `sys_clock_idle_exit` and `sys_clock_isr` (after `sys_clock_set_timeout(0, false)`) is 1-2 RTC ticks (62,5us - 125us), so quite significant. I was surprised about that, alignment to the tick boundary(100us in my system) may be the reason. I’m ready to move the announcement here, if you don’t see any problems.

Reply:
I've changed the way of announcement. I haven't encountered any issues with it.